fcrepo-http-api/src/main/java/org/fcrepo/http/api/FedoraReindex.java

```diff
@@ -0,0 +1,125 @@
+ * regarding copyright ownership.
+ *
+ * DuraSpace licenses this file to you under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except in
+ * compliance with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.fcrepo.http.api;
+
+import io.micrometer.core.annotation.Timed;
+import org.apache.http.HttpStatus;
+import org.fcrepo.kernel.api.identifiers.FedoraId;
+import org.fcrepo.kernel.api.services.ReindexService;
+import org.slf4j.Logger;
+import org.springframework.context.annotation.Scope;
+
+import javax.inject.Inject;
+import javax.ws.rs.BadRequestException;
+import javax.ws.rs.DELETE;
+import javax.ws.rs.GET;
+import javax.ws.rs.POST;
+import javax.ws.rs.PUT;
+import javax.ws.rs.Path;
+import javax.ws.rs.PathParam;
+import javax.ws.rs.Produces;
+import javax.ws.rs.core.Response;
+
+import static java.lang.String.format;
+import static javax.ws.rs.core.MediaType.APPLICATION_JSON;
+import static javax.ws.rs.core.Response.status;
+import static org.fcrepo.http.commons.domain.RDFMediaType.TEXT_PLAIN_WITH_CHARSET;
+import static org.slf4j.LoggerFactory.getLogger;
+
+/**
+ * @author dbernstein
+ * @since 12/01/20
+ */
+@Timed
+@Scope("request")
+@Path("/{path: (.+/)?}fcr:reindex")
+public class FedoraReindex extends FedoraBaseResource {
+
+    private static final Logger LOGGER = getLogger(FedoraReindex.class);
+
+    @Inject
+    private ReindexService reindexService;
+
+    @PathParam("path")
+    protected String externalPath;
+
+    /**
+     * Default JAX-RS entry point
+     */
+    public FedoraReindex() {
+        super();
+    }
+
+    /**
+     * Reindex a fedora resource.
+     *
+     * @return A response
+     */
+    @POST
+    @Produces({APPLICATION_JSON + ";qs=1.0",
+            TEXT_PLAIN_WITH_CHARSET})
+    public Response reindexObject() {
+        LOGGER.info("receiving reindex request for fedora_id = {}", externalPath);
+        try {
+            final var transaction = transaction();
+            final var id = FedoraId.create(externalPath);
+            if (doesResourceExist(transaction, id, true)) {
+                //TODO : remove this block once reindexing of existing resources is supported.
+                return status(HttpStatus.SC_CONFLICT, "Reindexing of existing resources is not currently supported. " +
+                        "Only resources that have not yet been indexed are allowed.").build();
+            } else {
+                try {
+                    final var baseId = id.asBaseId();
+                    transaction.lockResource(baseId);
```

Comment:
This is inconsistent with how locking is done in the other services. It will work, but, to me, having the kernel acquire a lock rather than the http layer seems more appropriate.

Reply:
@pwinckles :  I'm assuming the lock should also be released in the kernel layer as well (ie in the same routine as it is being done here in the http layer)?

Reply:
The way all of the other services work is that the kernel layer acquires the lock and the http layer initiates the transaction commit. The locks cannot be released before the transaction is committed, so so long as the transaction is committed in the http layer, the locks must also be released there.

If it is possible to push the commit down to the kernel, that does seem better to me, but I think we need to be consistent so that it's easy to reason about.